hal/src/thumbv6m/sercom/v2/uart.rs

```diff
@@ -0,0 +1,1930 @@
+    C::Word: PrimInt,
+    u16: AsPrimitive<C::Word>,
+    S: Sercom,
+    C::Pads: Rx,
+{
+    /// Read from the DATA register
+    ///
+    /// Reading from the data register directly is `unsafe`, because it will
+    /// clear the RXC flag, which could break assumptions made elsewhere in
+    /// this module.
+    #[inline]
+    pub unsafe fn read_data(&mut self) -> u16 {
+        self.sercom.usart().data.read().data().bits()
+    }
+
+    /// Clear interrupt status flags
+    ///
+    /// Setting the ERROR, SSL or TXC flag will clear the interrupt. Clearing
+    /// any flag will have no effect. This function has no effect on the DRE or
+    /// RXC flags.
+    ///
+    /// **Warning:** The implementation of of [`Write::flush`] waits on and
+    /// clears the `TXC` flag. Manually clearing this flag could cause it to
+    /// hang indefinitely.
+    #[inline]
+    pub fn clear_flags(&mut self, flags: Flags) {
+        unsafe {
+            self.sercom()
+                .usart()
+                .intflag
+                .write(|w| w.bits(flags.bits()))
+        };
+    }
+
+    /// Clear error status flags
+    ///
+    /// Setting a flag will clear the error. Clearing any flag will have no
+    /// effect.
+    #[inline]
+    pub fn clear_errors(&mut self, errors: Status) {
+        unsafe {
+            self.sercom()
+                .usart()
+                .status
+                .write(|w| w.bits(errors.bits()))
+        };
+    }
+
+    /// Read the interrupt flags
+    pub fn read_flags(&self) -> Flags {
+        <Self as Registers>::read_flags(self)
+    }
+
+    /// Read the status flags
+    #[inline]
+    pub fn read_status(&self) -> Status {
+        <Self as Registers>::read_status(self)
+    }
+
+    /// Flush the RX buffer and clear overflow errors
+    #[inline]
+    pub fn flush(&mut self) {
+        let usart = self.sercom.usart();
+
+        // TODO
+        // The datasheet states that disabling the receiver (RXEN) clears
+        // the RX buffer, and clears the BUFOVF, PERR and FERR bits.
+        // However, in practice, it seems like BUFOVF errors still pop
+        // up after a disable/enable cycle of the receiver, then immediately begin
+        // reading bytes from the DATA register.
+        //
+        // Is this a hardware bug???
+        /*
+        usart.ctrlb.modify(|_, w| w.rxen().clear_bit());
+        while usart.syncbusy.read().ctrlb().bit() || usart.ctrlb.read().rxen().bit_is_set() {}
+
+        usart.ctrlb.modify(|_, w| w.rxen().set_bit());
+        while usart.syncbusy.read().ctrlb().bit() || usart.ctrlb.read().rxen().bit_is_clear() {}
+        */
+
+        // Workaround is to read a few bytes to clear the RX buffer (3 bytes seems to do
+        // the trick), then manually clear the BUFOVF bit Clear rx buffer
+        for _ in 0..=2 {
+            let _data = usart.data.read();
+        }
+
+        // Clear buffer overflow error
+        self.clear_errors(Error::Overflow.into());
+    }
+}
```

Comment:
I wrote a `UartRx::flush` method to clear the rx buffer, as well as buffer overflow errors, in the event the receiver was enabled some time before data actually starts being read. However, I think I encountered a hardware bug (see commented code), leading to this inelegant workaround. 

Also, I couldn't find anywhere in the datasheet the actual length of the Rx buffer. After experimentation, it seems to be 3 bytes, as the buffer overflow errors pop up after reading the second byte. The BUFOVF bit gets set to true when the *next* word to be read has the error. I also don't know if the buffer for thumbv7em targets is also 3 bytes long.

Reply:
This is very weird. Did you ever follow up with an example sent to Microchip?

Reply:
I haven't yet. I might need to get a C example up.